src/Orleans/Lifecycle/ILifecycleObserver.cs

```diff
@@ -0,0 +1,10 @@
+﻿using System.Threading.Tasks;
+
+namespace Orleans
+{
+    public interface ILifecycleObserver
+    {
+        Task OnStart();
+        Task OnStop();
```

Comment:
It is done this way for error handling reasons.  Teardown should be opposite of startup and should only be done to the degree that startup succeeded.  If stages are started in order a, b, c , d, then they should be shutdown in d, c, b, a.  If during startup we fail at c, shutdown should start at c (not d).  Note that this is how grain activation lifecycle works now (most all lifecycles work this way), but current code has all of this hard coded in a way that is difficult to test, maintain, or extend.